Return error response for invalid JSON payloads

sloria opened the pull request:
Previously, invalid JSON payloads were allowed to pass silently.
With this change, a HTTPError with status code 400 and a payload of `{"json": ["Invalid JSON body."]}` is raised instead.

close #329

Note: Adds simplejson as a dependency on Python 2.

sloria committed:
Allow JSONDecodeError to be raised when invalid JSON is passed

Previously, invalid JSON payloads were allowed to pass silently.
With this change, the JSONDecodeError is raised.
Therefore, users are responsible for handling these errors.

Example with flask:

```python
@app.errorhandler(json.JSONDecodeError)
def handle_invalid_json(err):
    return jsonify(["Invalid JSON."]), 400
```

close #329

sloria committed:
Depend on simplejson and use it consistently

Ensures consistency of behavior across Python 2 and 3

lafrech commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,21 @@
+
+Other changes:
+
+* `simplejson <https://pypi.org/project/simplejson/>`_ is now a dependency.
```
Can you elaborate on this?

https://pypi.org/project/simplejson/

> simplejson is the externally maintained development version of the json library included with Python 2.6 and Python 3.0, but maintains backwards compatibility with Python 2.5.

I didn't test, but the README says it is included in 2.6+ and 3.0+. How come we need it in the first place?

sloria replied:
We need simplejson's backport of `json.JSONDecodeError`, which doesn't exist on Python 2.7.

lafrech replied:
Ah, OK.

Do you think we could make the requirement conditional, depending on the Python version?

Or add a comment to remember to remove it when dropping Python 2.x?

No big deal otherwise.

sloria replied:
Hm, conditionally requiring simplejson on Python 2 won't work OOB because

```python
In [1]: import simplejson

In [2]: import json

In [3]: simplejson.JSONDecodeError is json.JSONDecodeError
Out[3]: False
```

I suppose we could do something like

```python
# webargs/json.py

try:
    import simplejson as json
except ImportError:
    import json


# in parser modules
from webargs.core import parse_json
from webargs.json import json

...
try:
    parse_json(data)
except json.JSONDecodeError as err:
    if err.doc == "":
       ....
```

I'd be fine with that, I suppose.

sloria replied:
Actually, we could just import json from `webargs.core`:

```python
# in parser modules
from webargs.core import parse_json, json
...

try:
    parse_json(data)
except json.JSONDecodeError as err:
    if err.doc == "":
       ....
```

sloria replied:
OK, updated the code so that simplejson is only required on Python 2.

sloria committed:
Only require simplejson for Python 2.7

Import json with

```
from webargs.core import json
```

to ensure that we use the same implementation everywhere

lafrech commented:
I didn't inspect thoroughly so this is a bit of a lazy question, but wouldn't it be more in line with the validation error triggering a 422 if the parser aborted with 400 rather than raising an exception for the user to catch?

I find it a bit dodgy that we raise an exception that will trigger a 500 out of the box. Even with a major release and a warning.

I can live with that change. I'll just need to add a handler as you suggest, but is there a reason not to return a 400 error directly? The message could be personalized with a parser attribute if that's a concern. And the formatting can be done by catching the `HTTPError` just like with the 422. At least in the use case I'm familiar with (Flask + flask_rest_api).

sloria commented:
You raise a fair point. I think I went the 500 route for consistency with the current behavior of aiohttpparser and for configurability. But in retrospect, it's more user-friendly to raise an `HTTPError`. I'll try this out when I have some time.

sloria committed:
Return a 400 response instead of raising a JSONDecodeError

https://github.com/marshmallow-code/webargs/pull/334#issuecomment-450434221

sloria commented:
Rebased and modified so that a 400 response is returned with the payload:

```json
{"json": ["Invalid JSON body."]}
```

A `handle_invalid_json_error` hook was added to the relevant parsers to make this behavior configurable.

lafrech commented on examples/bottle_example.py:
```diff
@@ -58,6 +58,7 @@ def dateadd(value, addend, unit):
 # Return validation errors as JSON
+@error(400)
 @error(422)
 def error422(err):
```
Should this be renamed (remove 422 in the name)?

sloria replied:
Good catch. Fixed.

lafrech commented on setup.py:
```diff
@@ -1,8 +1,12 @@
 
 INSTALL_REQUIRES = ["marshmallow>=2.15.2"]
+if sys.version_info[0] < 3:
+    INSTALL_REQUIRES.append("simplejson")
```
Add a minimal version? "simplejson>=x.x.x"?

lafrech commented on tests/apps/bottle_app.py:
```diff
@@ -115,7 +115,8 @@ def echo_nested_many():
-@error(422)
+@app.error(400)
+@app.error(422)
 def handle_422(err):
```
Likewise, 422 in the name.

lafrech commented:
Looks good to me! :+1:

sloria committed:
Improve error handler names; error handling docs; min version of simplejson

h/t @lafrech

lafrech commented:
Great. This is good to go.
